File: 백트래킹/실버/실버3/15663_N과 M(9)/15663.py
import sys

def backtrack(N, M, sequence, numbers, visited):
    if len(sequence) == M:
        print(' '.join(map(str, sequence)))
        return
    
    last_used = None  # 마지막에 사용한 숫자 저장 변수
    for i in range(N):
        if not visited[i] and numbers[i] != last_used:
            sequence.append(numbers[i])
            visited[i] = True
            last_used = numbers[i]
            backtrack(N, M, sequence, numbers, visited)

            visited[i] = False
            sequence.pop()

def read_data(input_data):
    lines = [list(map(int, line.split())) for line in input_data.splitlines()]
    N, M = lines[0]
    numbers = sorted(lines[1])
    return N, M, numbers

if __name__ == '__main__':
    input_data = sys.stdin.read()
    N, M, numbers = read_data(input_data)
    visited = [False] * N
    backtrack(N, M, [], numbers, visited)

File: 백트래킹/실버/실버3/15663_N과 M(9)/15663.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int M = scanner.nextInt();
        int[] numbers = new int[N];
        
        for (int i = 0; i < N; i++) {
            numbers[i] = scanner.nextInt();
        }

        Arrays.sort(numbers);  // 입력된 숫자 정렬
        boolean[] visited = new boolean[N];
        List<Integer> sequence = new ArrayList<>();
        
        backtrack(N, M, sequence, numbers, visited);
        scanner.close();
    }

    public static void backtrack(int N, int M, List<Integer> sequence, int[] numbers, boolean[] visited) {
        if (sequence.size() == M) {
            for (int num : sequence) {
                System.out.print(num + " ");
            }
            System.out.println();
            return;
        }

        Integer lastUsed = null;
        for (int i = 0; i < N; i++) {
            if (!visited[i] && (lastUsed == null || numbers[i] != lastUsed)) {
                sequence.add(numbers[i]);
                visited[i] = true;
                lastUsed = numbers[i];

                backtrack(N, M, sequence, numbers, visited);

                visited[i] = false;
                sequence.remove(sequence.size() - 1);
            }
        }
    }
}
